### DISCUSSION TITLE
Current status overview (WIP)

### DISCUSSION THREAD

**Discussion by Bia10** [kikipoulet/SukiUI#156] [General]
This a summary of my notes and some discussion with @kikipoulet I had over discord.
This is WIP, anyone is free to expand/comment on this or even better feel free take up any item and contribute.

# Community

- Create some kind of quick communication server like Discord, as most of the time it is not suitable to put one line question as either issue or open discussion on it.

# Issues

## Buttons demo page

- Due to the way the buttons are styled on Buttons page, clicking on any button will force others to inherit the same states.
- The show xaml on the busy button will not work by simply pasting into playground as it has special reference to extensions.

## Property grid page

- Opening child form on PropertyGrid page actually overrides the the values in the output form from the child form even if nothing was entered.
- The input form and output form miss-match on present number of fields, therefore some input field changes are not translated to output form.
- The output form is not scrollable and henceforth required resizing of the window to see the bottom fields.

## Toast service

- Investigate how stable the internals are sometimes a rare bugs/exceptions occurs by pool failing to provide host for toast service.

## Misc

- Investigate the Avalonia diagnostics output. There's ton of warnings for some property of element being either null or whitespace.
- Refactor the codebase to utilize modern .net8/C#12 features and syntax.

# Missing features

#### Dispatcher Service

- Often time we call Avalonia dispatchers directly as such creating a repetitive ballast/scaffolding which should be abstracted somewhere away.

#### File/FolderPicker Service

- Basically a wrapper around the Avalonia window/topLevel.StorageProvider making it trivial to retrieve either a file/folders(s) or just their local/absolute path.

#### Toasts service

- It is planned to expand toasts to approach a style of Avalonia notifications and utilize the already present icons info/warn/error etc...

### Controls

- File/FolderPicker
- Docking system
- Carousel/Gallery
- Image viewer
- Signature/Draw pad
- Log/LogEvent viewer
- User/Avatar control
- FileSystem tree view

### Playground page

- Auto insert enclosing tags.
- Better autocompletion list, currently its not at first apparent how it works, which user events are used for selection which for insertion.
- Automatically select mouseover highlighted item as selected item for insertion.
- Since were now tracking the XML parsing exception we could highlight the actual place where the error occurred.
- Fill up the autocompletion list.
- Since there is no textBox header, you gotta find out all the input controls yourself, maybe add toolTips or textBox header?

## Misc

- Unit Testing
- CI/CD workflow

**Comment by kikipoulet** [comment]
@AuroraZiling  Just adding some resources about what's next, I think the next "big thing" missing is a docking system as it's one of the most asked thing on telegram channel for example. But it's quite difficult because : 

- Its design needs to fit with SukiUI, which is not easy as docking is usually old, compact and heavy design. 
- Shit it's not easy to code that thing

For now I just tried to imagine about what design could fit in, I just made that dumb UI model to imagine what a docking control for SukiUI could be : 

![Sans titre](https://github.com/kikipoulet/SukiUI/assets/19242427/71497c2d-54ad-4749-8272-b4e1c1983efb)

About the backend, we'll have to see if it's possible to create a custom one, doing a "minimal" docking control without advanced functionalities wouldn't be stupid too, or "cheating" by using the work of https://github.com/wieslawsoltes/Dock, maybe in an additional package, but risking to be nuget dependant to it.

**Reply by Insire** [reply]
If you want to support this, then the styles and what ever else is required for this to work, should be another nuget package IMO. I dont see any reason why depending on https://github.com/wieslawsoltes/Dock would be an issue. Its MIT licensed, so that should be fine

**Reply by AuroraZiling** [reply]
If it's feasible to depend on [dock](https://github.com/wieslawsoltes/Dock) and modify some styles to match the UI design consistency, this is a good solution.

btw, where's telegram channel or discord server, I want to join :)

**Reply by kikipoulet** [reply]
Thanks for your hints guys !

@AuroraZiling  you can join the telegram here : https://t.me/Avalonia

**Reply by kikipoulet** [reply]
I have promising result with Dock library : 

![docking](https://github.com/kikipoulet/SukiUI/assets/19242427/fd20c775-e24b-46f8-a6ea-f121fa081723)

I just hope it will be possible to make it a little smoother, maybe by digging CompositionAnimations
